### PR TITLE
JS: Recognize Express param value callback as RemoteFlowSource

### PR DESCRIPTION
**Pull request by asgerf**
A `param` route handler in Express takes the value of the parameter in the 4th argument as seen [here](https://gist.github.com/asgerf/f3da0abf24c90286c121034838e3f2e4). This PR adds it as a `RequestInputSource`.

### CODE REVIEW

**Comment by asgerf** [comment]
Thanks for pointing that out. Turns out we even misclassified the request and response parameters of these route handlers, due to them taking 4 parameters.

Updating this was more tricky than it had any right to be. PTAL.

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/js/express-param-handler_1588933876034) looks a bit swingy, will re-run the slowest.

File: javascript/ql/src/semmle/javascript/frameworks/ConnectExpressShared.qll [hunk 1]
```diff
@@ -7,24 +7,102 @@
...
+   * Holds if `fun` appears to match the given signature based on parameter naming.
+   */
+  private predicate matchesSignature(Function function, RouteHandlerSignature sig) {
+    function.getNumParameter() = sig.getArity() and
+    function.getParameter(sig.getParameterIndex("request")).getName() = ["req", "request"] and
+    function.getParameter(sig.getParameterIndex("response")).getName() = ["res", "response"] and
+    (
+      sig.has("next")
+      implies
+      function.getParameter(sig.getParameterIndex("next")).getName() = "next"
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/ConnectExpressShared.qll:61]: can we sneak in the dead change from https://github.com/github/codeql/pull/2229?
```suggestion
      function.getParameter(sig.getParameterIndex("next")).getName() = ["next", "cb"]
```
(fine if not)

**Comment by asgerf** [comment]
> Is this due to an optimizer limitation?

No, just formatting.

**Comment by asgerf** [comment]
[Re-run of slowest](https://git.semmle.com/asger/dist-compare-reports/tree/js/express-param-handler_1589234971429) is nowhere near as bad as in the original evaluation, but still a bit biased. I'll look at some tuple counts to see what's going.

**Comment by asgerf** [comment]
The use of `accesses` and `calls` had dubious join orderings, so I opted to switch to the `SourceNode`-based API instead. After rebasing on the `moduleImport`-caching PR this now seems to [perform well](https://git.semmle.com/asger/dist-compare-reports/tree/js/express-param-handler_1589541311928) but I'll run more evaluations over the weekend.

As a bonus, we now recognize `req.query[x]` as a `RemoteInputAccess`. 😅 

Hopefully that's the last feature-creep in this PR.

**Comment by asgerf** [comment]
[Security/nightly](https://git.semmle.com/asger/dist-compare-reports/tree/js/express-param-handler_1589614024890) looks good.

**Pull request merged**